README.md
# udp-service-discovery

Broadcasts Services via UDP (default every second via port 12345) and listens to them on the client.
If listenOnce(servicename) is used, the listening port will be unbound upon recieving a matching service. Client automatically tries to rebind if port is in use, so multiple instances can easyly run on the same host (they will bind one after another).

## Example


Server:

	var UDPServiceDiscovery = require('../dist/udp-service-discovery.js').UDPServiceDiscovery
    var udpDSCV = new UDPServiceDiscovery({port: 8000});

    // broadcasts name: 'my-server', ip: null (filled in by library), port: 4321, additional key/value {"protocol": "tcp"}
    udpDSCV.broadcast('my-server', null, 4321, {"protocol": "tcp", "other": 42});



Client:

	var UDPServiceDiscovery = require('../dist/udp-service-discovery.js').UDPServiceDiscovery
    var udpDSCV = new UDPServiceDiscovery({port: 8000});

    udpDSCV.listenOnce('my-server');

    udpDSCV.on('discovery', function (service) {
        // you would try to connect your client to service.host : service.port
        console.log(service);
    });


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

This project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [1.0.0] - 2017-11-21
### Changed
- converted the library to typescript
- updated examples to illustrate the latest changes

## [0.0.14] - 2017-11-19
### Fixed
- missing var declaration

## [0.0.13] - 2017-11-09
### Added
- added named export for typescript compatibility

## [0.0.12] - 2017-07-17
### Added
- added 'retryInterval' option
- implemented a delayed retry if the specified port is already in use

## [0.0.11] - 2017-07-13
### Changed
- dgram options to make the port reusable
- improved console log messages

## [0.0.10] - 2017-04-07
### Fixed
- service discovery could succeed if just the last element of a service of interest matched with the broadcast

## [0.0.9] - 2017-04-04
### Added
- ability to limit broadcasts to a certain number of repetitions

## [0.0.8] - 2017-02-22
### Changed
- introduced ES6 arrow functions
- cleaned up the code

## [0.0.7] - 2017-02-20
### Changed
- raised broadcast interval from 250ms to 1s




index.js
var util = require('util');
var dgram = require('dgram');
var EventEmitter = require('events').EventEmitter;
var Netmask = require('netmask').Netmask;

// named export for typescript users
exports.newUDPServiceDiscovery = UDPServiceDiscovery;
// default export for node
module.exports = UDPServiceDiscovery;

function UDPServiceDiscovery(opts) {

    if (!(this instanceof UDPServiceDiscovery)) {
        return new UDPServiceDiscovery(opts);
    }

    this.emit('statusChanged', this.status);
    this.status = "INITIALIZING";

    this.opts = opts || {};

    this.broadcasterPort = typeof this.opts.port === 'undefined' ? 12345 : this.opts.port;
    this.broadcasterAddress = typeof this.opts.address === 'undefined' ? null : this.opts.address;
    this.announceInterval = typeof this.opts.announceInterval === 'undefined' ? 1000 : this.opts.announceInterval;

    this.retryInterval = typeof this.opts.retryInterval === 'undefined' ? 999 : this.opts.retryInterval;

    this.serviceListenFor = null;
    this.serviceListenOnce = false;

    // Setup UDP socket
    this.socket = dgram.createSocket({type: 'udp4', reuseAddr: true});

    // Setup message handler
    this.socket.on('message', data => {
        var announcedService = JSONObjFromString(data);

        if (announcedService) {
            if (this.serviceListenFor) {
                var match = false;

                Object.keys(this.serviceListenFor).some(key => {
                    if (this.serviceListenFor[key] != announcedService[key]) {
                        match = false;
                        return true; // returning true on the first element that doesn't match to break the loop
                    } else {
                        match = true;
                    }
                });

                if (match) {
                    this.emit('discovery', announcedService);

                    if (this.serviceListenOnce) {
                        this.close();
                    }
                } else {
                    // no match
                }
            } else {
                this.emit('discovery', announcedService);

                if (this.serviceListenOnce) {
                    this.close();
                }
            }
        }
    });

    this.socket.on('error', e => {
        if (e.code === 'EADDRINUSE') {
            this._log('UDPServiceDiscovery address/port ' + e.address + ':' + e.port + ' in use, retrying in ' + this.retryInterval + ' ms');

            setTimeout(this.tryBinding.bind(this), this.retryInterval);
        } else {
            this._log('UDPServiceDiscovery SocketError: ' + e);
        }
    });

    this.socket.on('listening', () => {
        this.socket.setBroadcast(true);

        // this.socket.setMulticastLoopback(true);
        // this.socket.addMembership(state.address, state.host);

        var address = this.socket.address();

        if (this.listenOnce) {
            this._log('UDPServiceDiscovery listening (once) on ' + address.address + ':' + address.port);
        } else {
            this._log('UDPServiceDiscovery listening (forever) on ' + address.address + ':' + address.port);
        }
    });

    this.socket.on('close', () => {
        this._log('UDPServiceDiscovery Closing socket.');
    });
}

util.inherits(UDPServiceDiscovery, EventEmitter);

// name, ip, port, json object, broadcast-repeats (0 or any number)
UDPServiceDiscovery.prototype.broadcast = function broadcast(name, ip, port, msg) {
    var service = { host: null };
    var interval;
    var repeat = 0;
    var runs = 0;
    var netmask;
    var host;
    var broadcastAddress;
    var announceMessage;

    switch (arguments.length) {
        case 5:
            repeat = arguments[4];
        case 4:
            if (typeof arguments[3] === 'string') {
                service = JSONObjFromString(arguments[3]);
            } else {
                service = arguments[3];
            }
        case 3:
            service.port = arguments[2];
        case 2:
            service.host = arguments[1];
            service.name = arguments[0];
            break;
        case 1:
            service = JSONObjFromString(arguments[0]);
            break;
    }

    announceMessage = new Buffer(JSON.stringify(service));

    var announce = (() => {
        if (repeat != 0 && runs >= repeat) {
            clearInterval(interval);
            this._log('Finished ' + repeat + ' broadcasts.');
            return;
        }

        var mask = getLocalIPAndNetmask();
        if (mask.length > 0 && mask[0].length > 0) {
            netmask = mask[0][1];
            host = mask[0][0];
            broadcastAddress = getBroadcastAddress(this, host, netmask);

            if (host !== service.host) {
                service.host = host;
                announceMessage = new Buffer(JSON.stringify(service));
                this.emit('statusChanged', this.status);
            }

            if (host) {
                this.socket.send(announceMessage, 0, announceMessage.length, this.broadcasterPort, broadcastAddress, (err, bytes) => {
                    runs++;
                    if (err) {
                        this._log("UDP - Error announcing!", err);
                        throw err;
                    } else {
                        if (this.status !== 'BROADCASTING') {
                            this.status = "BROADCASTING";
                            this.emit('statusChanged', this.status);
                        }
                    }
                });
            }
        } else {
            // Not connected
            this.status = "NOT_CONNECTED";
            this.emit('statusChanged', this.status);
        }
    });

    // announce once per second
    interval = setInterval(announce, this.announceInterval);
};

UDPServiceDiscovery.prototype.listen = function listen(listenFor) {
    this.serviceListenOnce = false;

    if (listenFor) {
        if (typeof listenFor === 'string' && !JSONObjFromString(listenFor)) {
            this.serviceListenFor = {};
            this.serviceListenFor.name = listenFor;
        } else {
            if (typeof listenFor === 'string') {
                this.serviceListenFor = JSONObjFromString(listenFor);
            } else {
                this.serviceListenFor = listenFor;
            }
        }
    }

    this.tryBinding();
};

UDPServiceDiscovery.prototype.listenOnce = function listenOnce(listenFor) {
    this.serviceListenOnce = true;
    if (listenFor) {
        if (typeof listenFor === 'string' && !JSONObjFromString(listenFor)) {
            this.serviceListenFor = {};
            this.serviceListenFor.name = listenFor;
        } else {
            if (typeof listenFor === 'string') {
                this.serviceListenFor = JSONObjFromString(listenFor);
            } else {
                this.serviceListenFor = listenFor;
            }
        }
    }

    this.tryBinding();
};

UDPServiceDiscovery.prototype.tryBinding = function tryBinding() {
    this.socket.bind(this.broadcasterPort, this.broadcasterAddress);
};

UDPServiceDiscovery.prototype.close = function close() {
    this.socket.close();
};

UDPServiceDiscovery.prototype._log = function log() {
    if (this.opts.debug === true) {
        var args = Array.prototype.slice.call(arguments);
        console.log.apply(console, args);
    }
};

function JSONObjFromString(jsonString) {
    try {
        var o = JSON.parse(jsonString);

        // Handle non-exception-throwing cases:
        // Neither JSON.parse(false) or JSON.parse(1234) throw errors, hence the type-checking,
        // but... JSON.parse(null) returns null, and typeof null === "object",
        // so we must check for that, too. Thankfully, null is falsey, so this suffices:
        if (o && typeof o === 'object') {
            return o;
        }
    } catch (err) {
    }
    return null;
}

function getLocalIPAndNetmask() {
    var networkInterfaces = require('os').networkInterfaces();
    var matches = [];

    Object.keys(networkInterfaces).forEach(function (item) {
        networkInterfaces[item].forEach(function (address) {
            if (address.internal === false && address.family === 'IPv4') {
                matches.push([address.address, address.netmask]);
            }
        });
    });

    return matches;
}

function getBroadcastAddress(sender, ip, netmask) {
    sender._log(ip, netmask);

    var block = new Netmask(ip + "/" + netmask);

    return block.broadcast;
}


src/udp-service-discovery.ts
import * as dgram from 'dgram';
import { EventEmitter } from 'events';
import { Netmask } from 'netmask';


export interface UDPServiceDiscoveryOptions {
    port?: number;
    address?: string;
    announceInterval?: number;
    retryInterval?: number;
    debug?: boolean;
}

interface Service {
    host: string;
    port?: number;
    name?: string;
}

export class UDPServiceDiscovery extends EventEmitter {
    public get status() {
        return this._status;
    }

    private _opts: UDPServiceDiscoveryOptions;
    private _status = 'INITIALIZING';
    private _serviceListenFor = null;
    private _serviceListenOnce = false;
    private _socket: dgram.Socket;

    constructor(opts?: UDPServiceDiscoveryOptions) {
        super();

        this._opts = {
            port: 12345,
            address: null,
            announceInterval: 1000,
            retryInterval: 999,
            debug: false
        };

        if (typeof opts !== 'undefined') {
            Object.assign(this._opts, opts);
        }

        this._start();
    }

    broadcast(name: string, ip?: string, port?: number, msg?: any, repeat: number = 0) {
        let service: Service = { host: null };
        let interval;
        let runs = 0;
        let netmask;
        let host;
        let broadcastAddress;
        let announceMessage;

        switch (arguments.length) {
            case 5:
            case 4:
                if (typeof msg === 'string') {
                    service = this._getJSONObjFromString(msg);
                } else {
                    service = msg;
                }
            case 3:
                service.port = port;
            case 2:
                service.host = ip;
                service.name = name;
                break;
            case 1:
                service = this._getJSONObjFromString(arguments[0]);
                break;
        }

        announceMessage = new Buffer(JSON.stringify(service));

        const announce = (() => {
            if (repeat !== 0 && runs >= repeat) {
                clearInterval(interval);
                this._log('Finished ' + repeat + ' broadcasts.');
                return;
            }

            const mask = this._getLocalIPAndNetmask();
            if (mask.length > 0 && mask[0].length > 0) {
                netmask = mask[0][1];
                host = mask[0][0];
                broadcastAddress = this._getBroadcastAddress(this, host, netmask);

                if (host !== service.host) {
                    service.host = host;
                    announceMessage = new Buffer(JSON.stringify(service));
                    this.emit('statusChanged', this.status);
                }

                if (host) {
                    this._socket.send(announceMessage, 0, announceMessage.length, this._opts.port, broadcastAddress, (err, bytes) => {
                        runs++;
                        if (err) {
                            this._log('UDP - Error announcing!', err);
                            throw err;
                        } else {
                            if (this.status !== 'BROADCASTING') {
                                this._status = 'BROADCASTING';
                                this.emit('statusChanged', this.status);
                            }
                        }
                    });
                }
            } else {
                // Not connected
                this._status = 'NOT_CONNECTED';
                this.emit('statusChanged', this.status);
            }
        });

        // announce once per second
        interval = setInterval(announce, this._opts.announceInterval);
    }

    listen(listenFor) {
        this._serviceListenOnce = false;

        if (listenFor) {
            if (typeof listenFor === 'string' && !this._getJSONObjFromString(listenFor)) {
                this._serviceListenFor = {};
                this._serviceListenFor.name = listenFor;
            } else {
                if (typeof listenFor === 'string') {
                    this._serviceListenFor = this._getJSONObjFromString(listenFor);
                } else {
                    this._serviceListenFor = listenFor;
                }
            }
        }

        this.tryBinding();
    }

    listenOnce(listenFor) {
        this._serviceListenOnce = true;
        if (listenFor) {
            if (typeof listenFor === 'string' && !this._getJSONObjFromString(listenFor)) {
                this._serviceListenFor = {};
                this._serviceListenFor.name = listenFor;
            } else {
                if (typeof listenFor === 'string') {
                    this._serviceListenFor = this._getJSONObjFromString(listenFor);
                } else {
                    this._serviceListenFor = listenFor;
                }
            }
        }

        this.tryBinding();
    }

    tryBinding() {
        this._socket.bind(this._opts.port, this._opts.address);
    }

    private _start() {
        // Setup UDP socket

        this._socket = dgram.createSocket({type: 'udp4', reuseAddr: true});

        // Setup message handler
        this._socket.on('message', data => {
            const announcedService = this._getJSONObjFromString(data.toString());

            if (announcedService) {
                if (this._serviceListenFor) {
                    let match = false;

                    Object.keys(this._serviceListenFor).some(key => {
                        if (this._serviceListenFor[key] !== announcedService[key]) {
                            match = false;
                            return true; // returning true on the first element that doesn't match to break the loop
                        } else {
                            match = true;
                        }
                    });

                    if (match) {
                        this.emit('discovery', announcedService);

                        if (this._serviceListenOnce) {
                            this.close();
                        }
                    } else {
                        // no match
                    }
                } else {
                    this.emit('discovery', announcedService);

                    if (this._serviceListenOnce) {
                        this.close();
                    }
                }
            }
        });

        this._socket.on('error', (e: any) => {
            if (e.code === 'EADDRINUSE') {
                this._log('UDPServiceDiscovery address/port ' + e.address + ':' + e.port + ' in use, retrying in ' +
                    this._opts.retryInterval + ' ms');

                setTimeout(this.tryBinding.bind(this), this._opts.retryInterval);
            } else {
                this._log('UDPServiceDiscovery SocketError: ' + e);
            }
        });

        this._socket.on('listening', () => {
            this._socket.setBroadcast(true);

            const address = this._socket.address();

            if (this.listenOnce) {
                this._log('UDPServiceDiscovery listening (once) on ' + address.address + ':' + address.port);
            } else {
                this._log('UDPServiceDiscovery listening (forever) on ' + address.address + ':' + address.port);
            }
        });

        this._socket.on('close', () => {
            this._log('UDPServiceDiscovery Closing socket.');
        });
    }

    close() {
        this._socket.close();
    }

    private _log(...msg) {
        if (this._opts.debug) {
            console.log(...msg);
        }
    }

    private _getJSONObjFromString(jsonString: string) {
        try {
            const o = JSON.parse(jsonString);

            // Handle non-exception-throwing cases:
            // Neither JSON.parse(false) or JSON.parse(1234) throw errors, hence the type-checking,
            // but... JSON.parse(null) returns null, and typeof null === "object",
            // so we must check for that, too. Thankfully, null is falsey, so this suffices:
            if (o && typeof o === 'object') {
                return o;
            }
        } catch (err) {
        }
        return null;
    }

    private _getLocalIPAndNetmask() {
        const networkInterfaces = require('os').networkInterfaces();
        const matches = [];

        Object.keys(networkInterfaces).forEach(function (item) {
            networkInterfaces[item].forEach(function (address) {
                if (address.internal === false && address.family === 'IPv4') {
                    matches.push([address.address, address.netmask]);
                }
            });
        });

        return matches;
    }

    private _getBroadcastAddress(sender, ip, netmask) {
        sender._log(ip, netmask);
        const block = new Netmask(ip + '/' + netmask);
        return block.broadcast;
    }
}
